opm/simulators/flow/MICPContainer.cpp

```diff
@@ -0,0 +1,154 @@
+// -*- mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+// vi: set et ts=4 sw=4 sts=4:
+/*
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+
+#include <config.h>
+#include <opm/simulators/flow/MICPContainer.hpp>
+
+#include <opm/input/eclipse/Units/UnitSystem.hpp>
+
+#include <opm/output/data/Solution.hpp>
+
+#include <opm/simulators/flow/SolutionContainers.hpp>
+
+#include <algorithm>
+#include <array>
+#include <string>
+#include <tuple>
+
+namespace Opm {
+
+template<class Scalar>
+void MICPContainer<Scalar>::
+allocate(const unsigned bufferSize)
+{
+    cMicrobes_.resize(bufferSize, 0.0);
+    cOxygen_.resize(bufferSize, 0.0);
+    cUrea_.resize(bufferSize, 0.0);
+    cBiofilm_.resize(bufferSize, 0.0);
+    cCalcite_.resize(bufferSize, 0.0);
+
+    allocated_ = true;
+}
+
+template<class Scalar>
+void MICPContainer<Scalar>::
+assign(const unsigned globalDofIdx,
+       const Scalar microbialConcentration,
+       const Scalar oxygenConcentration,
+       const Scalar ureaConcentration,
+       const Scalar biofilmConcentration,
+       const Scalar calciteConcentration)
+{
+    cMicrobes_[globalDofIdx] = microbialConcentration;
+    cOxygen_[globalDofIdx] = oxygenConcentration;
+    cUrea_[globalDofIdx] = ureaConcentration;
+    cBiofilm_[globalDofIdx] = biofilmConcentration;
+    cCalcite_[globalDofIdx] = calciteConcentration;
+}
+
+template<class Scalar>
+MICPSolutionContainer<Scalar>
+MICPContainer<Scalar>::
+getSolution() const
+{
+    return {
+        cMicrobes_,
+        cOxygen_,
+        cUrea_,
+        cBiofilm_,
+        cCalcite_
+    };
+}
+
+template<class Scalar>
+void MICPContainer<Scalar>::
+outputRestart(data::Solution& sol) const
+{
+    if (!this->allocated_) {
+        return;
+    }
+
+    using DataEntry =
+        std::tuple<std::string, UnitSystem::measure, const std::vector<Scalar>&>;
+
+    auto doInsert = [&sol](const DataEntry& entry)
+    {
+        if (!std::get<2>(entry).empty()) {
+            sol.insert(std::get<std::string>(entry),
+                       std::get<UnitSystem::measure>(entry),
+                       std::move(std::get<2>(entry)),
```

Comment:
We're in a `const` member function and `get<2>(entry)` is a `const` object.  Applying `std::move()` to that will just copy and leave the original intact.  In the current master sources, we actually move the contents of the source array into the `Solution` object and that's why the `.empty()` calls are "everywhere".  It's not entirely clear to me that we preserve the full semantics by marking `outputRestart()` as `const` in this case.

Reply:
yeah that's an aspect of this i did not consider and compiler did not tell me about.

first, from a conceptual pov it is not really intuitive that an output function destroys the container. 

furthermore [the tuple entry](https://github.com/OPM/opm-simulators/blob/master/opm/simulators/flow/GenericOutputBlackoilModule.cpp#L412)  where I copied the code from always was a const vector<Scalar>&, the move has always been moot.

Reply:
> the move has always been moot.

Maybe not "always", but at least for a very long time.  In any case, we do have logic in place that depends on the `move()` actually moving, but as long as all output vectors continue to output once they've been written once that latent error would have been hidden.

Reply:
yeah, i just wanted to point out the problem. If these are the expected semantics I will implement them of course, but then we should fix the problem in the generic code as well.

Reply:
done